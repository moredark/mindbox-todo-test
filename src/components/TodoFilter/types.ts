import { FilterType } from "../../types";

export interface TodoFilterProps {
  activeFilter: FilterType;
  activeCount: number;
  onFilterChange: (filter: FilterType) => void;
  onClearCompleted: () => void;
  hasCompleted: boolean;
}
