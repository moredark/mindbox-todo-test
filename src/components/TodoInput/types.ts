export interface TodoInputProps {
  onAddTodo: (text: string) => void;
}