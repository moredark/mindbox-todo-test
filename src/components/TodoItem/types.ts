import { Todo } from '../../types';

export interface TodoItemProps {
  todo: Todo;
  onToggle: (id: string) => void;
  onDelete: (id: string) => void;
}
