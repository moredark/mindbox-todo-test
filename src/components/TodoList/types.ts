import { Todo } from "../../types";

export interface TodoListProps {
  todos: Todo[];
  toggleTodo: (id: string) => void;
  deleteTodo: (id: string) => void;
  isEmpty: boolean;
}
