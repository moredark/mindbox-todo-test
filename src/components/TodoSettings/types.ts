import { TodoSettings } from "../../types";

export interface TodoSettingsProps {
  settings: TodoSettings;
  onToggleAutoDelete: () => void;
}