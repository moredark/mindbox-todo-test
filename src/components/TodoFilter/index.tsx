import React from 'react';
import { Box, Button, Text } from 'grommet';
import { FilterButton } from '../common/FilterButton';
import { formatTaskCount } from '../../utils/formatters';
import { TodoFilterProps } from './types';

export const TodoFilter: React.FC<TodoFilterProps> = ({
    activeFilter,
    activeCount,
    onFilterChange,
    onClearCompleted,
    hasCompleted
}) => {
    return (
        <Box
            direction="row"
            align="center"
            justify="between"
            pad="small"
            background="background-contrast"
            round="small"
            margin={{ vertical: 'small' }}
            data-testid="todo-filter"
            border={{ color: 'dark-3', size: '1px' }}
        >
            <Text size="small" data-testid="todo-count" color="text">
                {formatTaskCount(activeCount)}
            </Text>

            <Box direction="row" gap="small">
                <FilterButton
                    label="All"
                    isActive={activeFilter === 'all'}
                    onClick={() => onFilterChange('all')}
                    testId="filter-all"
                />
                <FilterButton
                    label="Active"
                    isActive={activeFilter === 'active'}
                    onClick={() => onFilterChange('active')}
                    testId="filter-active"
                />
                <FilterButton
                    label="Completed"
                    isActive={activeFilter === 'completed'}
                    onClick={() => onFilterChange('completed')}
                    testId="filter-completed"
                />
            </Box>

            <Button
                label="Clear completed"
                onClick={onClearCompleted}
                disabled={!hasCompleted}
                size="small"
                color="accent-2"
                data-testid="clear-completed"
            />
        </Box>
    );
};
