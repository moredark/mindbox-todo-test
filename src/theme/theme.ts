import { ThemeType } from 'grommet';
import { deepMerge } from 'grommet/utils';
import { dark } from 'grommet/themes';

export const accentColor = '#7D4CDB';

export const customTheme: ThemeType = deepMerge(dark, {
  global: {
    colors: {
      background: '#121212',
      'background-back': '#1E1E1E',
      'background-front': '#2D2D2D',
      'background-contrast': '#3D3D3D',
      text: '#EEEEEE',
      'text-strong': '#FFFFFF',
      'dark-6': '#999999',
      'dark-5': '#BBBBBB',
      'accent-1': accentColor,
      brand: accentColor
    },
    focus: {
      border: {
        color: accentColor
      }
    }
  },
  button: {
    border: {
      radius: '4px'
    },
    primary: {
      color: accentColor
    }
  },
  checkBox: {
    hover: {
      border: {
        color: accentColor
      }
    },
    color: accentColor
  }
});
