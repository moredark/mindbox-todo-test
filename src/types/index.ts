export interface Todo {
  id: string;
  text: string;
  completed: boolean;
  createdAt: Date;
}

export interface TodoSettings {
  autoDeleteCompleted: boolean;
  // Другие настройки можно добавить здесь
}

export type FilterType = 'all' | 'active' | 'completed';
