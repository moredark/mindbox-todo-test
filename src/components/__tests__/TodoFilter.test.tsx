import { render, screen, fireEvent } from '@testing-library/react';
import { Grommet } from 'grommet';
import { TodoFilter } from '../TodoFilter';
import { customTheme } from '../../theme/theme';
import { formatTaskCount } from '../../utils/formatters';
import { FilterType } from '../../types';

jest.mock('../../utils/formatters', () => ({
  formatTaskCount: jest.fn((count) => `${count} items left`)
}));

describe('TodoFilter', () => {
  const defaultProps = {
    activeFilter: 'all' as FilterType,
    activeCount: 3,
    onFilterChange: jest.fn(),
    onClearCompleted: jest.fn(),
    hasCompleted: true
  };

  const renderComponent = (props = defaultProps) => {
    return render(
      <Grommet theme={customTheme}>
        <TodoFilter {...props} />
      </Grommet>
    );
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should display formatted task count', () => {
    renderComponent();
    expect(formatTaskCount).toHaveBeenCalledWith(3);
    expect(screen.getByTestId('todo-count')).toHaveTextContent('3 items left');
  });

  it('should call onFilterChange when filter button is clicked', () => {
    renderComponent();

    const activeButton = screen.getByTestId('filter-active');
    fireEvent.click(activeButton);

    expect(defaultProps.onFilterChange).toHaveBeenCalledWith('active');
  });

  it('should enable clear completed button when hasCompleted is true', () => {
    renderComponent();

    const clearButton = screen.getByTestId('clear-completed');
    expect(clearButton).not.toBeDisabled();
  });

  it('should disable clear completed button when hasCompleted is false', () => {
    renderComponent({
      ...defaultProps,
      hasCompleted: false
    });

    const clearButton = screen.getByTestId('clear-completed');
    expect(clearButton).toBeDisabled();
  });

  it('should call onClearCompleted when clear button is clicked', () => {
    renderComponent();

    const clearButton = screen.getByTestId('clear-completed');
    fireEvent.click(clearButton);

    expect(defaultProps.onClearCompleted).toHaveBeenCalledTimes(1);
  });

  describe('filter buttons', () => {
    it.each([
      ['all', 'filter-all'],
      ['active', 'filter-active'],
      ['completed', 'filter-completed']
    ] as const)('should handle %s filter click', (filter, testId) => {
      renderComponent();

      const button = screen.getByTestId(testId);
      fireEvent.click(button);

      expect(defaultProps.onFilterChange).toHaveBeenCalledWith(filter);
    });
  });
});
