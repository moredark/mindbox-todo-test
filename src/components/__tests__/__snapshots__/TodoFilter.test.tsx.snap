// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoFilter should match snapshot 1`] = `
<div>
  <div
    class="StyledGrommet-sc-19lkkz7-0 lioFhE"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 kMbDTu"
      data-testid="todo-filter"
    >
      <span
        class="StyledText-sc-1sadyjn-0 iXfSYi"
        data-testid="todo-count"
      >
        3 items left
      </span>
      <div
        class="StyledBox-sc-13pk1d4-0 bVWETU"
      >
        <button
          class="StyledButton-sc-323bzc-0 jfgVTo"
          data-testid="filter-all"
          style="background: none; border-bottom: 2px solid #7D4CDB; font-weight: bold;"
          type="button"
        >
          All
        </button>
        <div
          class="StyledBox__StyledBoxGap-sc-13pk1d4-1 hDDuS"
        />
        <button
          class="StyledButton-sc-323bzc-0 jrhiYA"
          data-testid="filter-active"
          style="background: none; font-weight: normal;"
          type="button"
        >
          Active
        </button>
        <div
          class="StyledBox__StyledBoxGap-sc-13pk1d4-1 hDDuS"
        />
        <button
          class="StyledButton-sc-323bzc-0 jrhiYA"
          data-testid="filter-completed"
          style="background: none; font-weight: normal;"
          type="button"
        >
          Completed
        </button>
      </div>
      <button
        class="StyledButton-sc-323bzc-0 kGGExe"
        data-testid="clear-completed"
        type="button"
      >
        Clear completed
      </button>
    </div>
  </div>
</div>
`;
